thirdparty/faiss/faiss/utils/onednn_utils.h

```diff
@@ -0,0 +1,216 @@
+    uint32_t xrow;
+    uint32_t xcol;
+    uint32_t yrow;
+    uint32_t ycol;
+    float* in_f32_1;
+    float* in_f32_2;
+    float* out_f32;
+    dnnl::memory::desc f32_md1;
+    dnnl::memory::desc f32_md2;
+    dnnl::memory::desc f32_dst_md2;
+    dnnl::memory f32_mem1;
+    dnnl::memory f32_mem2;
+    dnnl::memory f32_dst_mem;
+
+    dnnl::memory::desc bf16_md1;
+    dnnl::memory::desc bf16_md2;
+
+    dnnl::inner_product_forward::primitive_desc inner_product_pd;
+    dnnl::inner_product_forward inner_product_prim;
+
+    dnnl::memory bf16_mem1;
+
+    bool is_same(uint32_t xrow, uint32_t xcol, uint32_t yrow, uint32_t ycol,
+      float* in_f32_1, float* in_f32_2) {
+      return this->xrow == xrow && this->xcol == xcol && this->yrow == yrow && this->ycol == ycol && this->in_f32_2 == in_f32_2;
+    }
+
+    void init(uint32_t xrow, uint32_t xcol, uint32_t yrow, uint32_t ycol,
+		float* in_f32_1, float* in_f32_2) {
+        if (is_same(xrow, xcol, yrow, ycol, in_f32_1, in_f32_2)) {
+            if (this->in_f32_1 != in_f32_1) {
+                this->in_f32_1 = in_f32_1;
+                f32_mem1 = dnnl::memory(f32_md1, cpu_engine, in_f32_1);
+            }
+            return;
+        }
+
+        this->xrow = xrow;
+        this->xcol = xcol;
+        this->yrow = yrow;
+        this->ycol = ycol;
+        this->in_f32_1 = in_f32_1;
+        this->in_f32_2 = in_f32_2;
+
+        f32_md1 = dnnl::memory::desc({xrow, xcol}, dnnl::memory::data_type::f32, dnnl::memory::format_tag::ab);
+        f32_md2 = dnnl::memory::desc({yrow, ycol}, dnnl::memory::data_type::f32, dnnl::memory::format_tag::ab);
+        f32_dst_md2 = dnnl::memory::desc({xrow, yrow}, dnnl::memory::data_type::f32, dnnl::memory::format_tag::ab);
+
+        f32_mem1 = dnnl::memory(f32_md1, cpu_engine, in_f32_1);
+        f32_mem2 = dnnl::memory(f32_md2, cpu_engine, in_f32_2);
+        f32_dst_mem = dnnl::memory(f32_dst_md2, cpu_engine);
+
+        // inner memory bf16
+        bf16_md1 = dnnl::memory::desc({xrow, xcol}, dnnl::memory::data_type::bf16, dnnl::memory::format_tag::any);
+        bf16_md2 = dnnl::memory::desc({yrow, ycol}, dnnl::memory::data_type::bf16, dnnl::memory::format_tag::any);
+
+        inner_product_pd = dnnl::inner_product_forward::primitive_desc(
+              cpu_engine, dnnl::prop_kind::forward_training,
+              bf16_md1, bf16_md2, f32_dst_md2);
+
+        inner_product_prim = dnnl::inner_product_forward(inner_product_pd);
+
+        bf16_mem1 = dnnl::memory(inner_product_pd.src_desc(), cpu_engine);
+
+        // update state for new base data
+        if (this->in_f32_2 != in_f32_2)
+            BaseData::getState().store(BASE_DATA_STATE::MODIFIED);
+    }
+
+    void execut(float** out_f32) {
+        dnnl::reorder(f32_mem1, bf16_mem1).execute(engine_stream, f32_mem1, bf16_mem1);
+        BASE_DATA_STATE expected = BASE_DATA_STATE::MODIFIED;
+
+        if (BaseData::getState().compare_exchange_strong(expected, BASE_DATA_STATE::PREPARE)) {
```

Comment:
Plz CMIIW. In the first call, `expected` will be `BASE_DATA_STATE::MODIFIED` and changed into `BASE_DATA_STATE::PREPARE` in this line and return false. Then it will loop in line 196 forever

Reply:
The state is also designed for multi-thread scenario, the state change is INIT->MODIFIED -> PREPARE -> READY. When the first thread have finished the initialization, the other thread will get the state is READY, and then skip line 196.